pkg/ingester/ingester.go

```diff
@@ -1643,6 +1648,181 @@ func (i *Ingester) queryStreamChunks(ctx context.Context, db *userTSDB, from, th
+				Series: seriesInBatch,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			seriesInBatch = seriesInBatch[:0]
+		}
+	}
+
+	// Send any remaining series, and signal that there are no more.
+	err = client.SendQueryStream(stream, &client.QueryStreamResponse{
+		Series:              seriesInBatch,
+		IsEndOfSeriesStream: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure no error occurred while iterating the series set.
+	if err := ss.Err(); err != nil {
+		return nil, err
+	}
+
+	return allIterators, nil
+}
+
+func (i *Ingester) sendStreamingQueryChunks(allIterators []chunks.Iterator, stream client.Ingester_QueryStreamServer) (int, error) {
+	numSamples := 0
+
+	seriesInBatch := make([]client.QueryStreamSeriesChunks, 0, queryStreamBatchSize) // TODO: use a different value for queryStreamBatchSize?
+	batchSizeBytes := 0
+
+	// TODO: just use a slice for this? We'll never have more than queryStreamBatchSize series
+	chunkSlicePool := sync.Pool{
+		New: func() any {
+			// TODO: guess what size this will be based on time range?
+			return []client.Chunk{}
+		},
+	}
+
+	for seriesIdx, it := range allIterators {
+		seriesChunks := client.QueryStreamSeriesChunks{
+			Series: uint64(seriesIdx),
+			Chunks: chunkSlicePool.Get().([]client.Chunk),
+		}
+
+		for it.Next() {
+			meta := it.At()
+
+			// It is not guaranteed that chunk returned by iterator is populated.
+			// For now just return error. We could also try to figure out how to read the chunk.
+			if meta.Chunk == nil {
+				return 0, errors.Errorf("unfilled chunk returned from TSDB chunk querier")
+			}
+
+			ch := client.Chunk{
+				StartTimestampMs: meta.MinTime,
+				EndTimestampMs:   meta.MaxTime,
+				Data:             meta.Chunk.Bytes(),
+			}
+
+			switch meta.Chunk.Encoding() {
+			case chunkenc.EncXOR:
+				ch.Encoding = int32(chunk.PrometheusXorChunk)
+			case chunkenc.EncHistogram:
+				ch.Encoding = int32(chunk.PrometheusHistogramChunk)
+			case chunkenc.EncFloatHistogram:
+				ch.Encoding = int32(chunk.PrometheusFloatHistogramChunk)
+			default:
+				return 0, errors.Errorf("unknown chunk encoding from TSDB chunk querier: %v", meta.Chunk.Encoding())
+			}
+
+			seriesChunks.Chunks = append(seriesChunks.Chunks, ch)
+			numSamples += meta.Chunk.NumSamples()
+		}
+
+		msgSize := seriesChunks.Size()
+
+		// TODO: what values to use for queryStreamBatchSize and queryStreamBatchMessageSize?
+		if (batchSizeBytes > 0 && batchSizeBytes+msgSize > queryStreamBatchMessageSize) || len(seriesInBatch) >= queryStreamBatchSize {
+			// Adding this series to the batch would make it too big, flush the data and add it to new batch instead.
+			err := client.SendQueryStream(stream, &client.QueryStreamResponse{
+				SeriesChunks: seriesInBatch,
+			})
+			if err != nil {
+				return 0, err
+			}
+
+			batchSizeBytes = 0
+			seriesInBatch = seriesInBatch[:0]
+
+			for _, s := range seriesInBatch {
+				chunkSlicePool.Put(s.Chunks)
```

Comment:
I'm wondering if this is safe to do, or gRPC may still retain the `client.QueryStreamResponse`. This is a discussion me and Bryan had few times, and we've always found it quite difficult to answer this question.

**IIRC**, since this is a streaming gRPC, when we call `client.SendQueryStream()` we basically just send a frame. IIRC, this is **unsafe to do** because the `client.SendQueryStream()` should return once the message has been enqueued and doesn't guarantee it has been sent.

Reply:
Good point, I'll double-check.

Reply:
So this is not safe - sending the message itself is fine, but [this](https://github.com/grafana/mimir/blob/164fe86fb40d40bb76210adc7c6f0fed8d80b336/vendor/google.golang.org/grpc/stream.go#L1510) call to `LazyLog()` retains a copy of the message for an indefinite period of time. 

I'll remove it.

Reply:
Removed in b5fd50c08cb3279a08c7f2f4b1fc25679dbc427a